evaluate.py

```diff
@@ -0,0 +1,107 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+
+from pyro.contrib.forecast import backtest as pyro_backtest
+from pyro.ops.stats import crps_empirical, quantile
+
+from util import M5Data
+
+
+@torch.no_grad()
+def eval_mae(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_mae but does not average over batch dimensions.
+    """
+    pred = pred.median(0).values
+    return (pred - truth).abs().reshape(truth.shape[:-2] + (-1,)).mean(-1).cpu()
+
+
+@torch.no_grad()
+def eval_rmse(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_rmae but does not average over batch dimensions.
+    """
+    pred = pred.mean(0)
+    error = pred - truth
+    return (error * error).reshape(truth.shape[:-2] + (-1,)).mean(-1).sqrt().cpu()
+
+
+@torch.no_grad()
+def eval_crps(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_crps but does not average over batch dimensions.
+    """
+    return crps_empirical(pred, truth).reshape(truth.shape[:-2] + (-1,)).mean(-1).cpu()
+
+
+@torch.no_grad()
+def eval_pl(pred, truth):
+    """
+    Computes pinball loss over 9 quantiles 0.005, 0.165, 0.25, 0.5, 0.75, 0.835, 0.975, 0.995.
+    """
+    us = torch.tensor(M5Data.quantiles, dtype=pred.dtype, device=pred.device)
+    pred = quantile(pred, probs=us, dim=0)  # 9 x batch_shape x duration x D
+    error = pred - truth.unsqueeze(0)
+    us = us.reshape((-1,) + (1,) * (pred.dim() - 1)).expand(pred.shape)
+    error = (torch.where(error <= 0, us, 1 - us) * error).mean(0)  # mean accross all quantiles
+    return error.reshape(truth.shape[:-2] + (-1,)).mean(-1).cpu()
+
+
+DEFAULT_METRICS = {
+    "mae": eval_mae,
+    "rmse": eval_rmse,
+    "crps": eval_crps,
+    "pl": eval_pl,
+}
+
+
+def eval_wrmsse(raw_value, train_data, weight):
+    lag1 = train_data[..., 1:, :] - train_data[..., :-1, :]
+    # drop starting 0s
+    actual_length = (train_data.cumsum(-2) != 0).sum(-2)
```

Comment:
This is an issue. The scaling value is calculated based on train_data but stripping the leading 0.
```
Note that the denominator of RMSSE is computed only for the time-periods for which the examined product(s) are actively sold, i.e., the periods following the first non-zero demand observed for the series under evaluation.
```
We need to make sure that the smallest train_window has `actual_length >= 2`. I am not sure if we do it outside of backtesting and add a warning if we detect a violation. Or we just keep it as is.

Another issue is if the sale values are constant (unlikely to happen) during the train window, the `scale` (which is computed from `lag1`) can be 0. Hopefully, we won't face that issue.

Reply:
Can we simply specify `min_train_window` to `backtest()`? http://docs.pyro.ai/en/stable/contrib.forecast.html#pyro.contrib.forecast.evaluate.backtest

Reply:
Makes sense! I'll add a check for it.